kube-core/src/error_boundary.rs

```diff
@@ -0,0 +1,155 @@
+//! Types for isolating deserialization failures. See [`ErrorBoundary`].
+
+use std::{borrow::Cow, fmt::Display};
+
+use k8s_openapi::apimachinery::pkg::apis::meta::v1::ObjectMeta;
+use serde::Deserialize;
+use serde_value::DeserializerError;
+
+use crate::Resource;
+
+/// A wrapper type for T that lets deserializing the parent object succeed, even if the T is invalid.
+///
+/// For example, this can be used to still access valid objects from an `Api::list` call or `watcher`.
+// We can't implement Deserialize on Result<T, InvalidObject> directly, both because of the orphan rule and because
+// it would conflict with serde's blanket impl on Result<T, E>, even if E isn't Deserialize.
+#[derive(Debug, Clone)]
+pub struct ErrorBoundary<T>(pub Result<T, InvalidObject>);
+
+/// An object that failed to be deserialized by the [`ErrorBoundary`].
+#[derive(Debug, Clone)]
+pub struct InvalidObject {
+    // Should ideally be D::Error, but we don't know what type it has outside of Deserialize::deserialize()
+    // It *could* be Box<std::error::Error>, but we don't know that it is Send+Sync
```

Comment:
That is unfortunate. So we would basically have to expose the error type as a generic arg on InvalidObject and put bounds on the error type to be able to put it in here?

Reply:
Even then we wouldn't be able to enforce it at compile time, since we can't put further constraints on the trait methods. At most we could offer to downcast it as a runtime check.